### PR TITLE
[#13] Add delay between pull request updates

### LINKED ISSUE #13: Add timeout between repos
We are hitting GitHub rate limits (which cause a hard failure), so we should add a configurable cooldown between processing each repo, in order to avoid this rate limit.

### PR DESCRIPTION
**Pull request by sancho20021**
# Description

Add configurable cooldown duration between each
pull request update.

I decided to insert delay between each `submit_or_update_pull_request` and `submit_error_report` invocations, so this change also affects GitLab.

# Fixes #13

### CODE REVIEW

File: module.nix [hunk 1]
```diff
@@ -91,6 +91,11 @@ in {
           description = "Extra lines to add to pull request body";
           default = "";
         };
+        cooldown = mkOption {
+          type = int;
+          description = "Cooldown duration between updating pull requests";
```
Comment (balsoft) [inline module.nix:96]: Should probably specify that it's in milliseconds

Comment (sancho20021) [reply module.nix]: Fixed

**Review by balsoft — APPROVED** [review]
Don't forget to autosquash before merge

**Comment by sancho20021** [comment]
> Don't forget to autosquash before merge

Yes, of course. Do I need to wait for two approvals before merging (In this pull requests and others)?

**Comment by sancho20021** [comment]
Oops, fortunately I found some bugs:

**Comment by sancho20021** [comment]
Oops, fortunately I found some bugs:
```
INFO  update_daemon > Updating ssh://git@github.com/sancho20021/update-daemon-test
 INFO  update_daemon > 
 INFO  update_daemon > ssh://git@github.com/sancho20021/update-daemon-test:
nixpkgs 118ec238bf (2022-05-15) -> e0a42267f7 (2022-06-18)


thread 'tokio-runtime-worker' has overflowed its stack
fatal runtime error: stack overflow
Aborted (core dumped)
```
I'll try to figure out what caused that

**Comment by sancho20021** [comment]
I tried to shrink down the difference between `master` and this branch, and found out that after changing
[`submit_or_update_request(settings, handle, body, true).await?;`](https://github.com/serokell/update-daemon/blob/7b3cfcb5203997cc4fe007eb7dce8f5b05e95ba1/src/main.rs#L119)

to this:
```
async move { submit_or_update_request(settings, handle, body, true).await }.await?;
```
the error occurs when update-daemon submits a new PR:
```
thread 'tokio-runtime-worker' has overflowed its stack
fatal runtime error: stack overflow
Aborted (core dumped)
```
I don't understand what's going on, could anyone explain, why is it happening?

File: src/main.rs [hunk 2]
```diff
@@ -77,10 +81,33 @@ enum UpdateError {
...
+        let time_passed = Instant::now().duration_since(*locked_ts);
+
+        if time_passed >= delay {
+            let res = task.await;
+            *locked_ts = Instant::now();
+            return res;
+        } else {
+            tokio::time::sleep(delay - time_passed).await;
+        }
+    }
```
Comment (notgne2) [inline src/main.rs:103]: Does this loop ever actually... loop? Would something a bit like this not work instead?
```suggestion
    let time_passed = Instant::now().duration_since(*locked_ts);

    if time_passed < delay {
        tokio::time::sleep(delay - time_passed).await;
    }

    *locked_ts = Instant::now();
    }
```
then just removing the `task` argument and using this as something to wait on?

**Comment by notgne2** [comment]
@sancho20021 honestly, with enough experience of weird async runtime issues in runtime, I would expect it to be obvious to me what's going on here but actually it's not, I don't really know what the issue is. Though what would that change be for exactly?

File: src/main.rs (hunk 2 above)
Comment (sancho20021) [reply src/main.rs:103]: Yes, I also thought about removing the loop, so will do

**Comment by notgne2** [comment]
```
diff --git a/src/main.rs b/src/main.rs
index 36a0911..c7deb9d 100644
--- a/src/main.rs
+++ b/src/main.rs
@@ -4,7 +4,6 @@
 
 use std::path::Path;
 
-use futures::Future;
 use std::process::Command;
 use std::sync::Arc;
 use xdg::BaseDirectories;
@@ -81,26 +80,19 @@ enum UpdateError {
     RequestError(#[from] request::RequestError),
 }
 
-async fn perform_with_delay<T>(
-    task: T,
+async fn wait_for_delay(
     previous_ts: Arc<TMutex<Instant>>,
     delay: Duration,
-) -> T::Output
-where
-    T: Future,
-{
+) {
     let mut locked_ts = previous_ts.lock().await;
-    loop {
-        let time_passed = Instant::now().duration_since(*locked_ts);
-
-        if time_passed >= delay {
-            let res = task.await;
-            *locked_ts = Instant::now();
-            return res;
-        } else {
-            tokio::time::sleep(delay - time_passed).await;
-        }
+
+    let time_passed = Instant::now().duration_since(*locked_ts);
+
+    if time_passed < delay {
+        tokio::time::sleep(delay - time_passed).await;
     }
+
+    *locked_ts = Instant::now();
 }
 
 async fn update_repo(
@@ -141,23 +133,15 @@ async fn update_repo(
         repo.commit(&settings, diff.spaced())?;
         repo.push(&settings)?;
 
-        perform_with_delay(
-            submit_or_update_request(settings, handle, body, true),
-            previous_update,
-            delay,
-        )
-        .await?;
+        wait_for_delay(previous_update, delay).await;
+        submit_or_update_request(settings, handle, body, true).await?
     } else {
         info!("{}: Nothing to update", handle);
         if diff_default.len() > 0 {
             repo.push(&settings)?;
 
-            perform_with_delay(
-                submit_or_update_request(settings, handle, body, true),
-                previous_update,
-                delay,
-            )
-            .await?;
+            wait_for_delay(previous_update, delay).await;
+            submit_or_update_request(settings, handle, body, true).await?
         }
     }
     Ok(())
@@ -294,17 +278,14 @@ async fn main() {
                     Err(e) => {
                         error!("{}: {}", repo_longlived.handle, e);
                         let delay = (&settings as &UpdateSettings).cooldown;
-                        match perform_with_delay(
-                            request::submit_error_report(
-                                settings,
-                                repo.handle,
-                                format!(
-                                    "I tried updating flake.lock, but failed:\n\n```\n{}\n```",
-                                    e
-                                ),
+                        wait_for_delay(ts_copy2, delay).await;
+                        match request::submit_error_report(
+                            settings,
+                            repo.handle,
+                            format!(
+                                "I tried updating flake.lock, but failed:\n\n```\n{}\n```",
+                                e
                             ),
-                            ts_copy2,
-                            delay,
                         )
                         .await
                         {
```

I don't have testing set up, would this change still work right, and would it avoid the issue?

**Comment by notgne2** [comment]
nevermind, this wouldn't account for time spent actually running

**Comment by sancho20021** [comment]
> I don't have testing set up, would this change still work right, and would it avoid the issue?

Hmm, I tried just waiting in a function for 10ms, and this avoids the error! I think I will try to come up with a proper solution using your suggestion, and see what'll happen

**Pull request merged**